### PR TITLE
Rich text viewer | Fix for escape characters parsing in markdown

### PR DESCRIPTION
**Pull request by saikrishnan-ni**
# Pull Request

## 🤨 Rationale
- As markdown serialization from editor includes escaping markdown syntax characters and some special characters with a backslash `\`, viewer markdown parser needs to understand the escape backslashes to render the rich text output properly

## 👩‍💻 Implementation

- Added support for understanding escape backslashes in markdown parser logic

  For the below markdown
  image

  #### Expected Behavior
  - Should ignore escape backslashes
  image

  #### Actual Behavior
  - Renders escape backslashes 
  image

## 🧪 Testing

- Added unit test cases in `markdown-parser.spec.ts`

## ✅ Checklist

- [x] I have updated the project documentation to reflect my changes or determined no changes are needed.

### CODE REVIEW

File: change/@ni-nimble-components-54559267-8991-4065-b1bd-cb315fba44f2.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "patch",
+  "comment": "Added support for using escape character(backslash) in rich text viewer",
```
Comment (vikisekarNI) [inline change/@ni-nimble-components-54559267-8991-4065-b1bd-cb315fba44f2.json:3]:
```suggestion
  "comment": "Render special characters properly without unnecessary backslashes",
```

File: packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts [hunk 2]
```diff
@@ -232,6 +232,93 @@ describe('Markdown parser', () => {
                 'Bulleted list with bold and italics'
             ]);
         });
+
+        it('numbers with escape character should be parsed as paragraph', () => {
```
Comment (vikisekarNI) [inline packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts:236]: Is the expectation here to check whether special character `.` is rendered properly without backslash `\`? Test case title seems to not convey that. 

Consider putting all below added test cases into a new describe block `escape backslashes should be ignored while parsing` without crowding the existing describe block. By creating a new describe block, the test case title can be simple enough as below for each character you would like to test the parsing. 

```suggestion
        it('special character `.` should be parsed properly', () => {
```

File: packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts [hunk 3]
```diff
@@ -232,6 +232,93 @@ describe('Markdown parser', () => {
...
+                name: r`\## heading 2`,
+                tags: ['P'],
+                textContent: ['## heading 2']
+            },
+            {
+                name: r`\[link\](url)`,
+                tags: ['P'],
+                textContent: ['[link](url)']
+            }
+        ];
```
Comment (vikisekarNI) [inline packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts:303]: Do we also need to extend this test case to assert hyphens, sequence of stars, underscores, etc. 

image

File: packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts (hunk 2 above)
Comment (saikrishnan-ni) [reply packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts:236]: Yes, here the expectation is backslash should not be rendered and `1. ` should not make the item to be parsed as list item and retain as paragraph. Will move it under the separate describe for escape backslash ignoring

File: packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts (hunk 3 above)
Comment (saikrishnan-ni) [reply packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts:303]: Will add them

**Review by vikisekarNI — APPROVED** [review]

File: packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts [hunk 4]
```diff
@@ -234,6 +234,100 @@ describe('Markdown parser', () => {
         });
     });
 
+    describe('escape backslashes should be ignored while parsing', () => {
```
Comment (jattasNI) [inline packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts:237]: This looks like good coverage of invalid escape characters being ignored. But do we have coverage of valid characters being handled correctly? For example, that `\n` results in a line break or `\\` results in a backslash.

**Review by jattasNI — APPROVED** [review]

File: packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts (hunk 4 above)
Comment (saikrishnan-ni) [reply packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts:237]: We don't have a test separately to check for proper rendering of `\n`, `\\`. Will add them.

**Review by m-akinc — APPROVED** [review]

**Review by rajsite — APPROVED** [review]

File: change/@ni-nimble-components-54559267-8991-4065-b1bd-cb315fba44f2.json [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "patch",
+  "comment": "Render special characters properly without unnecessary backslashes",
```
[the remark this answers is not in the document]
Comment (aagash-ni) [reply change/@ni-nimble-components-54559267-8991-4065-b1bd-cb315fba44f2.json]: Updated

File: packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts [hunk 6]
```diff
@@ -234,6 +234,114 @@ describe('Markdown parser', () => {
...
+
+        it('\n backslash<n> should render a new line', () => {
+            const doc = RichTextMarkdownParser.parseMarkdownToDOM(r`\n`);
+
+            expect(getTagsFromElement(doc)).toEqual(['P']);
+            expect(getLeafContentsFromElement(doc)).toEqual([r`\n`]);
+        });
+
+        it('\\ double backslash should render a single backslash', () => {
+            const doc = RichTextMarkdownParser.parseMarkdownToDOM('\\');
```
[the remark this answers is not in the document]
Comment (aagash-ni) [reply packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts]: Updated

File: packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts [hunk 7]
```diff
@@ -234,6 +234,114 @@ describe('Markdown parser', () => {
...
+
+            expect(getTagsFromElement(doc)).toEqual(['P', 'P', 'P']);
+            expect(getLeafContentsFromElement(doc)).toEqual([
+                r`1.\ item 1`,
+                r`2. item 2`,
+                r`3.\item 3`
+            ]);
+        });
+
+        it('special character `.` should be parsed properly (bullet list test)', () => {
```
[the remark this answers is not in the document]
Comment (aagash-ni) [reply packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts]: In the bulleted list case, the special character here is `-`, So modified this by appending  `-` with a backslash  `\- item2` and asserted accordingly

**Pull request merged**